### ISSUE TITLE
Both GradientShap and LayerGradientShap do not allow to forward through an embedding layer

### ISSUE THREAD

**Issue by caesar-one** [pytorch/captum#771]
## 🐛 Bug

I was trying to implement an attribution function using GradientShap for a model from HuggingFace Transformer library however, since the input tensor (a long tensor) is corrupted with Gaussian noise (becoming a float tensor), it is not a permissible input in the first embedding layer. I have tried to use LayerGradientShap (as I did successfully with LayerIntegratedGradients), to calculate the attributions w.r.t. the output of the embedding layer, but it seems not to work in LayerGradiendShap.

## To Reproduce

```
        gs = LayerGradientShap(_model_forward, model_embedding_layer)
        attributions, delta = gs.attribute(
            inputs=(_input["input_ids"], _input["attention_mask"], _input["token_type_ids"]),
            baselines=(_baseline["input_ids"], _baseline["attention_mask"], _baseline["token_type_ids"]),
            return_convergence_delta=True,
            target=_target,
            **kwargs
        )
        return attributions
```

The model is a simple Bert model for sequence classification, and the input is similar to the following:
```
_baseline = {'input_ids': tensor([[101,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}
_input = {'input_ids': tensor([[ 101,  146,  112,  182, 4259, 1114, 1128, 1272,  146,  112,  182, 1177,
         6782,  106,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}
```
Where the first dimension is considered as batch size (=1).
However when it is forwarded to the model it becomes something like:
```
tensor([[ 101.0000,   94.2562,   72.3061,  117.4974, 2749.5693,  719.1877,
          728.2260,  821.1909,   94.2562,   72.3061,  117.4974,  759.8599,
         4378.3940,   68.4326,  102.0000],
        [ 101.0000,   47.4299,   36.3845,   59.1249, 1383.5873,  361.8963,
          366.4443,  413.2245,   47.4299,   36.3845,   59.1249,  382.3626,
         2203.2141,   34.4354,  102.0000],
        [ 101.0000,   26.5356,   20.3561,   33.0786,  774.0762,  202.4703,
          205.0148,  231.1869,   26.5356,   20.3561,   33.0786,  213.9206,
         1232.6332,   19.2656,  102.0000],
        [ 101.0000,  119.1220,   91.3812,  148.4945, 3474.9346,  908.9169,
          920.3395, 1037.8297,  119.1220,   91.3812,  148.4945,  960.3188,
         5533.4600,   86.4858,  102.0000],
        [ 101.0000,   94.6653,   72.6199,  118.0074, 2761.5034,  722.3092,
          731.3867,  824.7552,   94.6653,   72.6199,  118.0074,  763.1579,
         4397.3975,   68.7296,  102.0000]], requires_grad=True)
```

Notice that also the shape has changed (probably due to the default n_samples=5).
Since the input cannot be corrupted after the embedding layer, doesn't it make sense to use GradientShap with a model that has an embedding layer?

 - Captum version = 0.4.0
 - PyTorch Version = 1.8.1
 - Python version = 3.9.1

**Comment by bilalsal** [comment]
Hi @caesar-one ,

you are right, most Captum algorithms internally assume the input tensor is float not int.
We realize that this poses a problem when dealing with token indices (these cannot be perturbed into float values).

The response in https://github.com/pytorch/captum/issues/734#issuecomment-908581109 provides some recommendations on how to handle such cases, in particular, how to apply Captum to HuggingFace Transformer models.

Let me know if you need further assistance

**Comment by caesar-one** [comment]
Thank you very much for your answer @bilalsal !

Actually I already solved the issue for `IntegratedGradients` using `LayerIntegratedGradients`, exactly as the post suggests, however the same solution does not work for `GradientShap` with `LayerGradientShap`. 

Maybe because the nature of GradientShap is to directly corrupt the original input with a Gaussian noise (while in Integrated Gradients it only passes from the baseline to the input)? Indeed I'm not sure it is possible to use Gradient Shap on a model having an embedding layer (i.e. whose input cannot be corrupted by definition). 

I thought that using `LayerGradientShap` would have enabled to corrupt the input _after_ the embedding layer, as it happens with _LayerIntegratedGradients_ but it seems not to be the case, or am I doing something wrong? 

Thank you for your help and for this nice library.

**Comment by bilalsal** [comment]
Hi @caesar-one,

you are right about some a difference between `GradientShap` and `IntegratedGradients` causing the issues when using the `LayerGradientShap` on token indices.
It is probably caused by `GradientShap` adding noise by design to the provided input to create `n_samples` versions.

The solution I would recommend is to trying using [Interpretable Embeddings](https://captum.ai/api/utilities.html#interpretable-embeddings).
The following two tutorials demonstrate how this can be achieved with two models:
- [PyTorch VQA](https://captum.ai/tutorials/Multimodal_VQA_Interpret)
- [BERT](https://captum.ai/tutorials/Bert_SQUAD_Interpret2)

Hope this helps
